components/content/content-editor/controls/mention/autocomplete/controls.ts

```diff
@@ -0,0 +1,254 @@
+
+				if (matches && matches.length >= 2) {
+					const start = this.view.coordsAtPos(state.selection.from);
+
+					const box = this.$refs.container.offsetParent.getBoundingClientRect();
+
+					// If the text control is more than 50% down the page, open the control above ("inverted")
+					const relativeYPos = box.top / Screen.height;
+					if (relativeYPos >= 0.5) {
+						this.inverted = true;
+						this.top = 'auto';
+						this.bottom = start.top - box.top + 30 + 'px';
+					} else {
+						this.inverted = false;
+						this.top = start.top - box.top + 30 + 'px';
+						this.bottom = 'auto';
+					}
+
+					if (this.Screen.isXs) {
+						// On mobile, we want to position the element to the left border of the screen
+						this.left = '-' + box.left + 'px';
+					} else {
+						this.left = 'auto';
+					}
+
+					this.suggestion = matches[1];
+					this.updateSuggestions(this.suggestion);
+					this.visible = true;
+
+					return;
+				}
+			}
+		}
+		this.visible = false;
+	}
+
+	private async updateSuggestions(suggestion: string) {
+		this.users = MentionCache.getUsers(this.$route, suggestion).sort((a, b) =>
+			numberSort(b.match, a.match)
+		);
+
+		this.remoteSuggestions(suggestion);
+
+		if (this.inverted) {
+			this.selectedIndex = this.users.length - 1;
+
+			// If we are inverted, wait for the list to render and then scroll to the bottom.
+			await this.$nextTick();
+			if (this.$refs.list) {
+				this.$refs.list.scrollTop = this.$refs.list.scrollHeight;
+			}
+		} else {
+			this.selectedIndex = 0;
+		}
+	}
+
+	private remoteSuggestions(suggestion: string) {
+		// Stop any existing queued search timeout.
+		if (this.remoteSuggestionTimeout) {
+			clearTimeout(this.remoteSuggestionTimeout);
+			this.remoteSuggestionTimeout = undefined;
+			this.isLoading = false;
+		}
+
+		// If we already have the search results cached, use those and don't query backend.
+		if (this.searchResultsCache[suggestion]) {
+			const searchUsers = this.searchResultsCache[suggestion];
+			this.processSearchUsers(searchUsers);
+		} else {
+			// We set a timeout for 200ms here to not send right away when the user is typing fast.
+			this.remoteSuggestionTimeout = setTimeout(async () => {
+				this.isLoading = true;
+
+				const payload = await Api.sendRequest(
+					'/web/search/mention-suggestions?q=' + encodeURIComponent(suggestion),
+					undefined,
+					{ detach: true }
+				);
+
+				if (payload.users) {
+					const searchUsers = User.populate(payload.users);
+
+					// Add to cache
+					if (!this.searchResultsCache[suggestion]) {
+						this.searchResultsCache[suggestion] = searchUsers;
+					}
+
+					// Only process results if the currently active suggestion is still the one we initiated the search with.
+					if (suggestion === this.suggestion) {
+						this.processSearchUsers(searchUsers);
+					}
+				}
+
+				this.isLoading = false;
+			}, 200);
+		}
+	}
+
+	private processSearchUsers(searchUsers: User[]) {
```

Comment:
I don't think I like this approach of calculating matchability like this. We should be able to rely on order returned by the search as is. I understand this is some hack so we can merge the mentions from the document and the ones from remote searches?